### PR TITLE
IO forking

### PR DESCRIPTION
**Pull request by nomisRev**
Currently, we rely on the following snippet to fork starting an `IO`.
This PR removes that boilerplate throughout the code and will share the `Continuation`/`RestartCallback` instance throughout the loop instead of tossing it like before.

```
val a: suspend () -> A = {
      suspendCoroutine { ca: Continuation<A> ->
        IORunLoop.startCancelable(ioA, connA) { either: Either<Throwable, A> ->
          either.fold({ error ->
            ca.resumeWith(Result.failure(error))
          }, { a ->
            ca.resumeWith(Result.success(a))
          })
        }
      }
    }

```

I don't think can optimize this further since we want to guarantee forking in all concurrency operators (in the case the CC is capable of it).

### CODE REVIEW

**Comment by nomisRev** [comment]
@pakoito this change changes the threading semantics a little bit. Before we the result in the receiver `CoroutineContext` but now it finishes on the last seen `ctx`. Here `"6"` instead of `"all"`.

```
"parallel mapping is done in the expected CoroutineContext" {
  fun makePar(num: Long) =
    IO(newSingleThreadContext("$num")) {
      // Sleep according to my number
      Thread.sleep(num * 20)
      num
    }
  val result =
    newSingleThreadContext("all").parMapN(
      makePar(6), IO.just(1L), makePar(4), IO.defer { IO.just(2L) }, makePar(5), IO { 3L }) { _, _, _, _, _, _ ->
      Thread.currentThread().name
    }.unsafeRunSync()
  result shouldBe "all"
}
```

File: modules/effects/arrow-effects-data/src/test/kotlin/arrow/effects/IOTest.kt [hunk 1]
```diff
@@ -301,11 +299,11 @@ class IOTest : UnitSpec() {
 
       val result =
         newSingleThreadContext("all").parMapN(
-          makePar(6), IO.just(1L), makePar(4), IO.defer { IO.just(2L) }, makePar(5), IO { 3L }) { _, _, _, _, _, _ ->
+          makePar(6), just(1L), makePar(4), IO.defer { just(2L) }, makePar(5), IO { 3L }) { _, _, _, _, _, _ ->
           Thread.currentThread().name
         }.unsafeRunSync()
 
-      result shouldBe "all"
+      result shouldBe "6"
```
Comment (pakoito) [inline modules/effects/arrow-effects-data/src/test/kotlin/arrow/effects/IOTest.kt]: Why 6 and not any other?

File: modules/effects/arrow-effects-data/src/main/kotlin/arrow/effects/IO.kt [hunk 2]
```diff
@@ -123,6 +126,39 @@ sealed class IO<out A> : IOOf<A> {
...
+   *
+   * ```kotlin:ank:playground
+   * import arrow.effects.*
+   * import arrow.effects.extensions.io.async.async
+   * import arrow.effects.extensions.io.monad.binding
+   * import kotlinx.coroutines.Dispatchers
+   *
+   * fun main(args: Array<String>) {
+   *   //sampleStart
+   *   binding {
```
Comment (raulraja) [inline modules/effects/arrow-effects-data/src/main/kotlin/arrow/effects/IO.kt]: should use `fx`

File: modules/effects/arrow-effects-data/src/main/kotlin/arrow/effects/IO.kt [hunk 3]
```diff
@@ -123,6 +126,39 @@ sealed class IO<out A> : IOOf<A> {
...
+   *     val promise = Promise.uncancelable<ForIO, Int>(IO.async()).bind()
```
Comment (raulraja) [inline modules/effects/arrow-effects-data/src/main/kotlin/arrow/effects/IO.kt]: `!`

File: modules/effects/arrow-effects-data/src/main/kotlin/arrow/effects/IO.kt [hunk 4]
```diff
@@ -123,6 +126,39 @@ sealed class IO<out A> : IOOf<A> {
...
+   *     val fiber = promise.get().fix().startFiber(Dispatchers.Default).bind()
```
Comment (raulraja) [inline modules/effects/arrow-effects-data/src/main/kotlin/arrow/effects/IO.kt]: Should use our dispatchers

File: modules/effects/arrow-effects-data/src/test/kotlin/arrow/effects/MVarTest.kt [hunk 5]
```diff
@@ -78,9 +78,9 @@ class MVarTest : UnitSpec() {
         binding {
           val av = mvar.empty<Int>().bind()
 
-          val f1 = av.put(10).startFiber(Dispatchers.Default).bind()
-          val f2 = av.put(20).startFiber(Dispatchers.Default).bind()
-          val f3 = av.put(30).startFiber(Dispatchers.Default).bind()
+          val f1 = av.put(10).fix().startFiber(Dispatchers.Default).bind()
```
Comment (raulraja) [inline modules/effects/arrow-effects-data/src/test/kotlin/arrow/effects/MVarTest.kt]: not sure why we need `fix()` if `startFiber` is for all `Kind<F, A> : Concurrent<F>`

File: modules/effects/arrow-effects-data/src/test/kotlin/arrow/effects/MVarTest.kt [hunk 6]
```diff
@@ -78,9 +78,9 @@ class MVarTest : UnitSpec() {
...
+          val f2 = av.put(20).fix().startFiber(Dispatchers.Default).bind()
```
Comment (raulraja) [inline modules/effects/arrow-effects-data/src/test/kotlin/arrow/effects/MVarTest.kt]: Ideally, when we change these we also change them to use `!`

File: modules/effects/arrow-effects-data/src/test/kotlin/arrow/effects/IOTest.kt (hunk 1 above)
Comment (nomisRev) [reply modules/effects/arrow-effects-data/src/test/kotlin/arrow/effects/IOTest.kt]: 6 always finishes last in this test.

```
fun makePar(num: Long) =
  IO(newSingleThreadContext("$num")) {
    // Sleep according to my number
    Thread.sleep(num * 20)
    num
  }
```

Comment (pakoito) [reply modules/effects/arrow-effects-data/src/test/kotlin/arrow/effects/IOTest.kt]: Put it on a comment IMO.

**Comment by nomisRev** [comment]
@raulraja I wanted to postpone to do this in 1 all in one go during #1435 fx on type classes PR.

Rewriting felt like a waste since the fx API is going to change, and we can rewrite them with the full power of the DSL (without passing CC / Concurrent everywhere).

**Pull request merged**